content/rust/docker/rust-docker-arm.md

```diff
@@ -12,3 +12,365 @@ tags:
+
+# Setup a Rust Project
+
+First we setup an example project. To do this we can use `cargo`:
+
+```
+$ cargo init hello-world
+```
+
+This will create a folder named `hello-world` which contains the project. It consists of a project file `Cargo.toml` and a source file `main.rs` in the `src` subdirectory. The source file contains code which simply prints `Hello, world!` in the console:
+
+{{< highlight text "linenos=table,hl_lines=99" >}}
+fn main() {
+    println!("Hello, world!");
+}
+{{< / highlight >}}
+
+Use the `cargo` tool to compile our program to a standalone binary file, and then execute the binary to start it:
+
+```
+$ cargo build
+$ ./target/debug/hello-world
+```
+
+The application should build without problems and display the message when run.
+
+# Cross Compiling a Rust Project
+
+Our goal is now to build this project for the ARM platform. Rust can build executables for many platforms. To find out which platforms are available use the following command:
+
+```
+$ rustc --print target-list
+```
+
+More informations about the supported platforms can be found [here](https://forge.rust-lang.org/release/platform-support.html). This tutorial uses `aarch64-unknown-linux-gnu` to compile the project for the ARM64 architecture.
+
+To compile a project for a different platform two things are required: the Rust Standard Library for this platform and a compatible linker. The standard library can be obtained with the `rustup` tool:
+
+```
+$ rustup target add aarch64-unknown-linux-gnu
+```
+
+It is a little bit harder to find a compatible linker to use. For `aarch64-unknown-linux-gnu` we have to install the `g++-aarch64-linux-gnu libc6-dev-arm64-cross` packages.
+
+```
+$ apt-get install g++-aarch64-linux-gnu libc6-dev-arm64-cross
+```
+
+To tell `cargo` to build for a specific platform we must pass the platform as a `--target` parameter and set the linker to use in an environment variable. The variable name has the format `CARGO_TARGET_<plaform>_LINKER` where `<platform>` is the platform in UPPERCASE letters.
+
+```
+$ export CARGO_TARGET_AARCH64_UNKNOWN_LINUX_GNU_LINKER=aarch64-linux-gnu-gcc
+$ cargo build --target=aarch64-unknown-linux-gnu
+//or
+$ CARGO_TARGET_AARCH64_UNKNOWN_LINUX_GNU_LINKER=aarch64-linux-gnu-gcc cargo build --target=aarch64-unknown-linux-gnu
+```
+
+The build process will create a new folder `aarch64-unknown-linux-gnu` in the `target` folder which contains the compiled binary file. If run it exits with an error message because the current system has not the correct platform:
+
+```
+hello-world: cannot execute binary file: Exec format error
+```
+
+To simplify this setup process we can use the project [cross](https://github.com/rust-embedded/cross). `cross` comes with a set of Docker images prepared for cross compilation.
+
+```
+$ cargo install cross
+$ cross build --target aarch64-unknown-linux-gnu
+```
+
+{{< alert info >}}
```

Comment:
although cross cannot be used in docker-in-docker, I was wondering if we could use the cross docker images directly instead of building our own? for example:

```
steps:
- name: build
  image: rustembedded/cross:aarch64-unknown-linux-gnu-0.1.16
  commands:
  - cargo build
```
```

Reply:
Thats a good question, I didn't think about that. I will add a note about this. I will not remove the compilation part because I think it is a good lesson to see how to setup a build environment in a docker container.

Reply:
perhaps creating a custom build image is a separate tutorial that we link to?  If it is possible to use `rustembedded/cross:aarch64-unknown-linux-gnu-0.1.16` (or equivalent) I think it would be preferred since it lowers the barrier to entry for new users.

Reply:
actually, can we just use the default rust image with drone and specify the arch in the platform?  Docker will use the manifest to pull down the correct rust image for the host machine architecture.

```
platform:
  arch: arm64

steps:
- name: build
  image: rust:debian
  commands:
  - cargo build
```

a `docker pull rust:debian` should download the arm64 image from Dockerhub

Reply:
But it isn't cross compilation anymore if we use a machine with the correct platform?

Reply:
thanks for looking into this. I think we are all set